File: 프로그래머스/1/250137. ［PCCP 기출문제］ 1번 ／ 붕대 감기/［PCCP 기출문제］ 1번 ／ 붕대 감기.py
def solution(bandage, health, attacks):
    t, x, y = bandage
    max_health = health

    attack_idx = 0
    success = 0

    last_time = attacks[-1][0]

    for time in range(1, last_time + 1):

        # 공격 받는 시간
        if time == attacks[attack_idx][0]:
            damage = attacks[attack_idx][1]

            health -= damage
            if health <= 0:
                return -1

            success = 0
            attack_idx += 1

        # 회복 시간
        else:
            success += 1

            health += x

            if success == t:
                health += y
                success = 0

            health = min(health, max_health)

    return health

File: 프로그래머스/1/250137. ［PCCP 기출문제］ 1번 ／ 붕대 감기/［PCCP 기출문제］ 1번 ／ 붕대 감기.java
class Solution {
    public int solution(int[] bandage, int health, int[][] attacks) {
        
        //보너스 받기위한 시전시간
        int t=bandage[0];
        //회복량
        int x=bandage[1];
        //보너스 회복량
        int y=bandage[2];
        //최대 회복량
        int max_health=health;
        
        int last_time=attacks[attacks.length-1][0];
        int attack_index=0;
        int continued_index=0;
        
        for(int time=1; time<=last_time; time++){
            if(attacks[attack_index][0]==time){
                health-=attacks[attack_index][1];
                if(health<=0){
                    return -1;
                }
                attack_index++;
                continued_index=0;
            }
            else{
                health+=x;
    
                continued_index++;
                if(continued_index==t){
                    health+=y;
                    continued_index=0;
                }
                
                if(health>=max_health){
                    health=max_health;
                }
            }
        
        }
        
        return health;
    }
}